src/transformers/generation/utils.py

```diff
@@ -3647,48 +3520,69 @@ def _assisted_decoding(
 
             # Store scores, attentions and hidden_states when required
             # Assistant: modified to append one tuple element per token, as in the other generation methods.
-            if return_dict_in_generate:
-                newly_added_length = n_matches + 1
-                if output_scores:
-                    scores += tuple(new_logits[:, i, :] for i in range(newly_added_length))
-                if output_logits:
-                    raw_logits += tuple(next_token_logits[:, i, :] for i in range(newly_added_length))
-
-                newly_added_length = new_cur_len if is_first_iteration else newly_added_length
-                if output_attentions:
-                    if self.config.is_encoder_decoder:
-                        cross_attentions = _split_model_outputs(
-                            cross_attentions, outputs.cross_attentions, cur_len, newly_added_length
-                        )
-                        decoder_attentions = _split_model_outputs(
-                            decoder_attentions,
-                            outputs.decoder_attentions,
-                            cur_len,
-                            newly_added_length,
-                            is_decoder_attention=True,
-                        )
-                    # some (V)LLMs have hard requirement on SDPA and thus never return attn
-                    elif outputs.attentions[0] is not None:
-                        decoder_attentions = _split_model_outputs(
-                            decoder_attentions,
+            # ... after computing new_logits, next_token_logits, and valid_tokens
+            newly_added = n_matches + 1
```

Comment:
imo this logic should go into `self._accumulate_generate_output`, so that future custom techniques can benefit from the splitting logic as well.

or better yet, a separation function to accumulate several outputs!